source/single_async_manual_reset_event.cpp

```diff
@@ -0,0 +1,182 @@
+  // succeeds then we've been successful and we'll need to invoke set_done after
+  // destroying the callback, which we'll do in start_or_wait.  We can't enter
+  // state 3 from here because we're being cancelled before calling
+  // start_or_wait, which is the only way to update state_ to opAddr, so if the
+  // CAS from nullptr fails then we must have transitioned to state 2a.
+  //
+  // In state 2a, we'll do nothing, start_or_wait will see that state_ is
+  // signalled, and immediately invoke complete, whereupon we'll check whether
+  // the stop_token has had stop requested (which it has), and we'll invoke
+  // set_done instead of set_value.
+  //
+  // In state 2b, the signalling thread is racing with the cancelling thread,
+  // which will be synchronized in complete when we destroy the stop callback.
+  // We'll then check whether the stop token has had stop requested, find that
+  // it has, and invoke set_done so we'll do nothing here.
+  //
+  // In state 3, our first CAS will succeed, signalling in state_ that we've
+  // been cancelled.  No future signal attempt will invoke complete so we'll
+  // call unifex::set_done ourselves after destroying the callback.
+
+  const auto opAddr = to_addr(op);
+  const auto signalledState = to_addr(this);
+  const auto cancelledState = signalledState + 1;
+
+  auto expectedState = opAddr;
+
+  if (state_.compare_exchange_strong(
+      expectedState,
+      cancelledState,
+      std::memory_order_acq_rel)) {
+    // state 3
+
+    // leave it in the no-waiter state
+    clear_cancelled_state(state_, cancelledState);
+
+    complete_operation(opAddr);
+    return;
+  }
+
+  if (expectedState == to_addr(nullptr)) {
+    // state 1
+    if (state_.compare_exchange_strong(
+        expectedState,
+        cancelledState,
+        std::memory_order_acq_rel)) {
+      // do nothing and let start_or_wait invoke complete
+      return;
+    }
+  }
+
+  if (expectedState == signalledState) {
+    // state 2 (a or b) or failed retry in state 1
+    return;
+  }
+  else {
+    std::terminate();
+  }
+}
+
+
+void single_async_manual_reset_event::start_or_wait(_op_base* op) noexcept {
+  const std::uintptr_t opAddr = to_addr(op);
+  const std::uintptr_t signalledState = to_addr(this);
+  const std::uintptr_t cancelledState = signalledState + 1;
+
+  std::uintptr_t expectedState = to_addr(nullptr);
+
+  if (state_.compare_exchange_strong(
+      expectedState,
+      opAddr,
+      std::memory_order_acq_rel)) {
+    // no one was waiting and it wasn't signalled, so now we're waiting
+    return;
+  }
+  else if (expectedState == signalledState) {
+    // already signalled; don't wait
+    complete_operation(opAddr);
+    return;
+  }
+  else if (expectedState == cancelledState) {
+    // cancelled before starting; clean up then complete
+
+    // leave it in the no-waiter state
+    clear_cancelled_state(state_, cancelledState);
```

Comment:
This is problematic as it might race with another thread calling `set()` which could cause the "signalled" state of the event to be lost.